tests/unit/Framework/Constraint/CountTest.php

```diff
@@ -142,4 +142,98 @@ public function testCountTraversable(): void
         $this->assertNotInstanceOf(\IteratorAggregate::class, $datePeriod);
         $this->assertTrue($countConstraint->evaluate($datePeriod, '', true));
     }
+
+    public function testCountingNonRewindableIteratorWithDifferentCount(): void
+    {
+        $this->assertCountFails(
+            new Count(2),
+            new \NoRewindIterator(new \ArrayIterator([1, 2, 3]))
+        );
+    }
+
+    public function testCountingNonRewindableIteratorWithSameCount(): void
+    {
+        $this->assertCountSucceeds(
+            new Count(2),
+            new \NoRewindIterator(new \ArrayIterator([1, 2]))
+        );
+    }
+
+    public function assertCountFails(Count $count, iterable $iterable): void
+    {
+        $this->assertFalse($count->evaluate($iterable, '', true));
+    }
+
+    public function assertCountSucceeds(Count $count, iterable $iterable): void
+    {
+        $this->assertTrue($count->evaluate($iterable, '', true));
+    }
+
+    public function testCountingDifferentGeneratorsSequentiallyWorks(): void
+    {
+        $count = new Count(2);
+
+        $generatorMaker             = new \TestGeneratorMaker();
+        $generatorWithThreeElements = $generatorMaker->create(['a', 'b', 'c']);
+        $generatorWithTwoElements   = $generatorMaker->create(['a', 'b']);
+
+        $this->assertCountFails($count, $generatorWithThreeElements);
+        $this->assertCountSucceeds($count, $generatorWithTwoElements);
+    }
+
+    public function testFailureMessageIsCorrectForGenerators(): void
+    {
+        $this->assertEvaluateFailsWithMessage(
+            new Count(1),
+            (new \TestGeneratorMaker())->create(['a', 'b']),
+            'actual size 2 matches expected size 1'
+        );
+    }
+
+    public function testFailureMessageIsCorrectForNonRewindableIterators(): void
+    {
+        $this->assertEvaluateFailsWithMessage(
+            new Count(1),
+            new \NoRewindIterator(new \ArrayIterator([1, 2, 3])),
+            'actual size 3 matches expected size 1'
+        );
+    }
+
+    public function testFailureMessageIsCorrectForTraversablesThatAreNotIteratorsOrIteratorAggregates(): void
+    {
+        $this->assertEvaluateFailsWithMessage(
+            new Count(1),
+            $this->newPdoStatementWithThreeValues(),
+            'actual size 3 matches expected size 1'
+        );
+    }
+
+    private function assertEvaluateFailsWithMessage(Count $count, iterable $iterable, string $message): void
+    {
+        try {
+            $count->evaluate($iterable);
+        } catch (ExpectationFailedException $e) {
+            $this->assertContains(
+                $message,
+                TestFailure::exceptionToString($e)
+            );
+
+            return;
+        }
+
+        $this->fail('An exception should have been thrown');
+    }
+
+    private function newPdoStatementWithThreeValues(): \Traversable
+    {
+        if (!\class_exists('\PDO')) {
```

Comment:
```php
        if (!\extension_loaded('pdo') || !\in_array('sqlite', \PDO::getAvailableDrivers())) {
```
according to https://github.com/sebastianbergmann/phpunit/blob/be6564ea49b96b9776c53f627a49683ec2687ce5/tests/end-to-end/regression/GitHub/1351.phpt